### PR TITLE
DynamoDB Enhanced Client: Added support for attribute level custom up…

### PR DESCRIPTION
**Pull request by bmaizels**
…date behaviors

Implements suggestion in https://github.com/aws/aws-sdk-java-v2/issues/1822

## Description
Copied from README.md update:

### Changing update behavior of attributes
It is possible to customize the update behavior as applicable to individual attributes when an 'update' operation is
performed (e.g. UpdateItem or an update within TransactWriteItems).

For example, say like you wanted to store a 'created on' timestamp on your record, but only wanted its value to be
written if there is no existing value for the attribute stored in the database then you would use the 
WRITE_IF_NOT_EXISTS update behavior. Here is an example using a bean:

```java
@DynamoDbBean
public class Customer extends GenericRecord {
    private String id;
    private Instant createdOn;

    @DynamoDbPartitionKey
    public String getId() { return this.id; }
    public void setId(String id) { this.name = id; }

    @DynamoDbUpdateBehavior(UpdateBehavior.WRITE_IF_NOT_EXISTS)
    public Instant getCreatedOn() { return this.createdOn; }    
    public void setCreatedOn(Instant createdOn) { this.createdOn = createdOn; }
}
```

Same example using a static table schema:

```java
static final TableSchema<Customer> CUSTOMER_TABLE_SCHEMA =
     TableSchema.builder(Customer.class)
       .newItemSupplier(Customer::new)
       .addAttribute(String.class, a -> a.name("id")
                                         .getter(Customer::getId)
                                         .setter(Customer::setId)
                                         .tags(primaryPartitionKey()))
       .addAttribute(Instant.class, a -> a.name("createdOn")
                                          .getter(Customer::getCreatedOn)
                                          .setter(Customer::setCreatedOn)
                                          .tags(updateBehavior(UpdateBehavior.WRITE_IF_NOT_EXISTS)))
       .build();
```

### CODE REVIEW

File: services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/models/RecordWithUpdateBehaviors.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    @DynamoDbUpdateBehavior(WRITE_IF_NOT_EXISTS)
+    public Instant getCreatedOn() {
+        return createdOn;
+    }
```
Comment (millems) [inline services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/models/RecordWithUpdateBehaviors.java:46]: I'm confused about the use of this. Would you actually use it for "created date" if it doesn't work for putItem (only for updateItem)? Wouldn't this be `firstUpdatedDate`, which seems kind of useless?

Comment (bmaizels) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/models/RecordWithUpdateBehaviors.java:46]: The use-case, as it was put to me, was to avoid branching logic and always do a uniform update. It also allows schema evolution so they can update the record and write a value with a new attribute that didn't exist in the schema before only if it hadn't been set before in the database. I imagine it's the same rationale that lead to DDB implementing if_not_exists in their update expressions at all, this is really just a 'plumbing it through' exercise.

Comment (millems) [reply services-custom/dynamodb-enhanced/src/test/java/software/amazon/awssdk/enhanced/dynamodb/functionaltests/models/RecordWithUpdateBehaviors.java:46]: I think it's probably more useful at the request level than the schema level, but that's a different feature request, I suppose.

**Review by millems — APPROVED** [review]

**Comment by takarabt** [comment]
Hi @bmaizels,

Are you planning to add the possibility to ignore null values when updating a record?

I think this is a very common use case. For example, if I want to change only the description for a given record using the DynamoDBMapper (sdk v1) I could use the object {"id": "123", "description": "My new description"}, Using the same object with the DynamoDB Enhanced client, the attributes which are not included on the object will be overwritten to null.

Basically we need something similar to UPDATE_SKIP_NULL_ATTRIBUTES -> https://aws.amazon.com/blogs/developer/using-the-savebehavior-configuration-for-the-dynamodbmapper/

For more flexibility, it's probably better to add this as a configuration on the client level.

Regards,

**Comment by bmaizels** [comment]
@takarabt does https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/enhanced/dynamodb/model/UpdateItemEnhancedRequest.html#ignoreNulls-- cover your use case?

**Comment by takarabt** [comment]
@bmaizels this is exactly what we are looking for, thank you.

**Comment by takarabt** [comment]
@bmaizels yes this is what we used, after investigation I figured out that the problem is related the usage of primitive type "int" it's set by default to 0 and not to null.

Thank you.
